File: Problems & Solutions/T/Total Decoding Messages/27 January 2023/Total Decoding Messages.py
class Solution:
	def CountWays(self, str):
		if str[0] == '0' :
				return 0
			
		n = len(str)
		mod = 1000000007
		
		dp = [0 for i in range(n+1)]
		
		dp[0] = dp[1] = 1
		
		for i in range(1,n):
		    if str[i] == '0' and str[i-1] > '2' :
		        return 0
		       
		
		for i in range(2,n+1):
			if str[i-1] > '0' :
			    dp[i] = dp[i-1]
		
			if str[i-2] == '1' or (str[i-2] == '2' and str[i-1] < '7') :
			    dp[i] +=dp[i-2]
			
			dp[i] %= mod
		
		return dp[n]

#{ 
 # Driver Code Starts
#Initial Template for Python 3

import sys
sys.setrecursionlimit(10**6)
if __name__ == '__main__':
	T=int(input())
	for i in range(T):
		str = input()
		ob = Solution()
		ans = ob.CountWays(str)
		print(ans)

# } Driver Code Ends

File: Problems & Solutions/T/Total Decoding Messages/27 January 2023/Total Decoding Messages.java
//{ Driver Code Starts
//Initial Template for Java

import java.util.*;
import java.lang.*;
import java.io.*;
class GFG
{
    public static void main(String[] args) throws IOException
    {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        PrintWriter out=new PrintWriter(System.out);
        int T = Integer.parseInt(br.readLine().trim());
        while(T-->0)
        {
            String str = br.readLine().trim();
            Solution ob = new Solution();
            int ans = ob.CountWays(str);
            out.println(ans);
        }
        out.close();
    }
}

// } Driver Code Ends


//User function Template for Java

class Solution
{
    static int mod=1000000007;
    public int CountWays(String str)
    {
        long dp1 = 1, dp2 = 0;
        int n = str.length();
        for(int i=n-1; i>=0; i--) {
            
            long dp = str.charAt(i) == '0' ? 0 : dp1;
            
            if(i<n-1 && (str.charAt(i) == '1' || str.charAt(i) == '2' && str.charAt(i+1) < '7'))
                dp = (dp+dp2) % mod;
            dp2 = dp1;
            dp1 = dp;
        }
        
        return (int)(dp1 % mod);
    }
}
